### DISCUSSION TITLE
Embedding Qualtrics Survey in jsPsych

### DISCUSSION THREAD

**Discussion by SHogenboom** [jspsych/jsPsych#1889] [Q&A] [answered]
Hi,

Has anyone perhaps fiddled around - and succeeded - with embedding Qualtrics surveys within jsPsych experiments? 
I have see some posts about people redirecting from a Qualtrics survey to a jsPsych experiment, but not as an embedded survey. 

My idea:
1. Use jsPsych to manipulate survey flow, conditions, loops etc
2. Use Qualtrics to create a pretty & functional survey (with many question types not yet integrated in jsPsych plugins) 

Looking for tips specifically on:
* How to pass data from the jsPsych environment to the embedded survey (e.g., condition number)
* How to validate the end of the embedded survey -> i.e., start the next trial.

**Comment by becky-gilbert** [comment, marked as answer]
Hi @SHogenboom, Qualtrics surveys have to be presented via a Qualtrics server, which is why it's not really possible to embed them in a jsPsych experiment. If you don't want to redirect between websites, then I think the best you can do is to present the Qualtrics survey inside of an [iframe](https://www.w3schools.com/html/html_iframe.asp) (which is just presenting one webpage inside of another one). 

One way to do this might be to use the jsPsych [external-html plugin](https://www.jspsych.org/plugins/jspsych-external-html/), but I'm not sure if that plugin allows you to present a remotely-hosted page, or if it's just for HTML files in the local experiment directory. Another option might be to use the jsPsych [call-function plugin](https://www.jspsych.org/plugins/jspsych-call-function/) to put the iframe on the page. You would need to set the `async` parameter to `true` so that the call-function trial waits until the participant is finished with the survey before moving on with your jsPsych experiment. Again: I haven't tried this so don't know if it works! If those plugins don't work, then you could always create your own plugin for this purpose.

To pass information from jsPsych to Qualtrics (e.g. participant and condition numbers), you can use URL query parameters in your Qualtrics link and save them to Embedded Data (see [Qualtrics docs here](https://www.qualtrics.com/support/survey-platform/survey-module/survey-flow/standard-elements/passing-information-through-query-strings/)).

I'm not totally sure how you can detect the end of a survey that's running in an iframe in order to end the Qualtrics 'trial' and move on with the jsPsych experiment. I came across a few options on [this stackoverflow post](https://stackoverflow.com/questions/16589306/get-data-from-iframe) and [this blog post](https://blog.pavey.dev/how-to-easily-send-and-receive-data-with-an-iframe-ckgd69r9f0886pas1h68s0vav).

Let us know if you figure out a solution for this and/or need further guidance.

**Reply by SHogenboom** [reply]
@becky-gilbert thanks for your reply and your ideas! 

For now I will not actually pursue your suggestions as my supervisor since pointed out - rightly so - that creating new plugins would be better from an open science perspective. So rather than trying to integrate Qualtrics (which will also take time and effort) I'll continue with a new plugin which allows for multiple form types within the same question (e.g., an other option that takes text input).

**Reply by becky-gilbert** [reply]
We totally agree WRT creating and sharing plugins! :smiley: We're happy to help with the plugin development if we can - feel free to post your code and questions on the Discussion forum. 

And by the way, we are planning to revamp the survey plugins in the near future so that they're more flexible - see Issue #1207. But this is at least a few months away, so maybe not soon enough for you. If you'd like to contribute to the new 'official' survey plugins, feel free to comment on the Issue post with ideas and code, and/or submit a pull request.

**Comment by Max-Lovell** [comment]
@SHogenboom you might find this website useful: https://kywch.github.io/jsPsych-in-Qualtrics/

**Reply by SHogenboom** [reply]
Thanks for the suggestion - but it's not a solution unfortunately. The link you posted discusses using jsPsych experiments in Qualtrics surveys. This is possible because you can embed a jsPsych experiment. As Becky pointed out, it is not possible to do the reverse (Qualtrics in jsPsych).

**Reply by Max-Lovell** [reply]
Ah, sorry for the misunderstanding. 

I see you're going down a different route but I had a thought based on something I did recently so I'll share anyway in case someone comes across this and is stumped. I did embed jsPsych in a iFrame here: https://github.com/Max-Lovell/dots_metacognition_task/tree/main/qualtrics_iFrame so I assume the reverse method could work. Maybe using the stim: function() option in a jsPsych canvas keyboard response, you could appending the qualtrics body to the document body. You'd need to find and collapse the jsPsych body and HTML doc to 0 width and height. at the end of the qualtrics survey some JavaScript code in a text/graphic qualtrics question that finds and removes the qualtrics iFrame in the Qualtrics.SurveyEngine.addOnUnload(function(){}); could work. You could communicate between the two with a jQuery.Ajax() function or URLs with query strings.
